src/gui/processing/qgsprocessingmeshdatasetwidget.cpp

```diff
@@ -0,0 +1,568 @@
+  if ( mMeshLayer && mParam )
+  {
+    int scalarDatasetGroup = mMeshLayer->rendererSettings().activeScalarDatasetGroup();
+    int vectorDatasetGroup = mMeshLayer->rendererSettings().activeVectorDatasetGroup();
+
+    if ( scalarDatasetGroup >= 0 && mMeshLayer->datasetGroupMetadata( scalarDatasetGroup ).dataType() == mParam->dataType() )
+      options.append( scalarDatasetGroup );
+
+    if ( vectorDatasetGroup >= 0
+         && mMeshLayer->datasetGroupMetadata( vectorDatasetGroup ).dataType() == mParam->dataType()
+         && vectorDatasetGroup != scalarDatasetGroup )
+      options.append( vectorDatasetGroup );
+  }
+
+  setValue( options );
+}
+
+QgsProcessingMeshDatasetGroupsWidgetWrapper::QgsProcessingMeshDatasetGroupsWidgetWrapper( const QgsProcessingParameterDefinition *parameter, QgsProcessingGui::WidgetType type, QWidget *parent ):
+  QgsAbstractProcessingParameterWidgetWrapper( parameter, type, parent )
+{}
+
+QString QgsProcessingMeshDatasetGroupsWidgetWrapper::parameterType() const
+{
+  return QgsProcessingParameterMeshDatasetGroups::typeName();
+}
+
+QgsAbstractProcessingParameterWidgetWrapper *QgsProcessingMeshDatasetGroupsWidgetWrapper::createWidgetWrapper( const QgsProcessingParameterDefinition *parameter, QgsProcessingGui::WidgetType type )
+{
+  return new QgsProcessingMeshDatasetGroupsWidgetWrapper( parameter, type );
+}
+
+void QgsProcessingMeshDatasetGroupsWidgetWrapper::postInitialize( const QList<QgsAbstractProcessingParameterWidgetWrapper *> &wrappers )
+{
+  QgsAbstractProcessingParameterWidgetWrapper::postInitialize( wrappers ); //necessary here?
+  switch ( type() )
+  {
+    case QgsProcessingGui::Standard:
+    case QgsProcessingGui::Batch:
+    {
+      for ( const QgsAbstractProcessingParameterWidgetWrapper *wrapper : wrappers )
+      {
+        if ( wrapper->parameterDefinition()->name() == static_cast< const QgsProcessingParameterMeshDatasetGroups * >( parameterDefinition() )->meshLayerParameterName() )
+        {
+          setMeshLayerWrapperValue( wrapper );
+          connect( wrapper, &QgsAbstractProcessingParameterWidgetWrapper::widgetValueHasChanged, this, [ = ]
+          {
+            setMeshLayerWrapperValue( wrapper );
+          } );
+          break;
+        }
+      }
+      break;
+    }
+
+    case QgsProcessingGui::Modeler:
+      break;
+  }
+}
+
+void QgsProcessingMeshDatasetGroupsWidgetWrapper::setMeshLayerWrapperValue( const QgsAbstractProcessingParameterWidgetWrapper *wrapper )
+{
+  // evaluate value to layer
+  QgsProcessingContext *context = nullptr;
+  std::unique_ptr< QgsProcessingContext > tmpContext;
+  if ( mProcessingContextGenerator )
+    context = mProcessingContextGenerator->processingContext();
+
+  if ( !context )
+  {
+    tmpContext = qgis::make_unique< QgsProcessingContext >();
+    context = tmpContext.get();
+  }
+
+  QgsMeshLayer *meshLayer = QgsProcessingParameters::parameterAsMeshLayer( wrapper->parameterDefinition(), wrapper->parameterValue(), *context );
```

Comment:
See logic in QgsProcessingFieldWidgetWrapper::setParentLayerWrapperValue -- you need to take ownership of the layer here.

Reply:
If I understand well the logic in QgsProcessingFieldWidgetWrapper::setParentLayerWrapperValue, we need ownership because if context come from `context=tmpContext.get();` the layer will be destruct at the end of this scope.

If the context come from mProcessingContextGenerator, we also take it and if needed in other scope, the layer will produce again another one that will be in the temporary store.

If I am not wrong, for mesh layer, there will be several instance of the same layer, one more for each wrapper (here the datasetGroup and the datasetTime wrappers). What I don't like is the memory consuming of this approach which the mesh layer, as each layer have all the data (mesh , dataset,).

Does the life time of the context in the context generator is greater that the life time of the wrapper?
In that case, maybe taking the ownership will be useful only if the context generator provide a nullptr?

Reply:
Finally the approach has been changed as this logic is not very adapted for mesh layer.
Now the logic is 

- when setting the mesh layer to the widget, if the mesh layer is in the project, the widget keeps a pointer to the layer to update value if needed (mesh layer can change : current active dataset groups, extra dataset groups, virtual dataset groups, ...)

- If the mesh layer is not in the project, the widget do not keep a pointer but extract all what is needed, here as the layer is not in the project, it could not change.